content/go/concepts/interfaces/interfaces.md

```diff
@@ -0,0 +1,262 @@
+
+An interface with no methods is called an empty interface.
+
+```go
+type interfaceName interface {}
+```
+
+The tradeoff of using an empty interface is the loss of type safety. As there are no methods, it's possible to pass any type to a function that takes an empty interface as an argument.
+
+### Example
+
+The `Println` function in the `fmt` package takes an empty interface as an argument.
+
+Below is the go source code for the `fmt.Println` function:
+
+```go
+// Println formats using the default formats for its operands and writes to standard output.
+// Spaces are always added between operands and a newline is appended.
+// It returns the number of bytes written and any write error encountered.
+func Println(a ...interface{}) (n int, err error) {
+    return Fprintln(os.Stdout, a...)
+}
+```
+
+> ##Note:\*\* due to the empty interface, we can pass any type to the `Println` function.
+
+## Embedding an Interface
+
+An interface in go can be embedded in other interfaces or structs.
+
+```go
+package main
+
+import "fmt"
+
+// any type that need to implement marriage should also implement all methods of dating
+type marriage interface {
+    dating
+    love()
+    arranged()
+}
+
+type dating interface {
+    compatibility()
+}
+
+type john struct {
+    age int
+}
+
+func (a john) love() {
+    fmt.Println("love with john")
+}
+
+func (a john) arranged() {
+    fmt.Println("arranged marriage with john")
+}
+func (a john) compatibility() {
+    fmt.Println("compatible with john")
+}
+
+type parents struct {
+    //named marriage interface
+    m     marriage
+    quote string
+}
+
+type lovers struct {
+    //unnamed marriage interface
+    marriage
+    quote string
+}
+
+func main() {
+    a := john{age: 24}
+    p := parents{quote: "I saw that you were perfect, and so I loved you. Then I saw that you were not perfect and I loved you even more", m: a}
+
+    //p.love()
+    //as this has named marriage interface m calling directly with .m would raise compiler error
+    /*-----------------------ERROR-----*---------------------------*/
+    // ./prog.go:39:4: p.love undefined (type parents has no field or method love)
+
+    p.m.love()
+    p.m.compatibility()
+    p.m.arranged()
+
+    l := lovers{quote: "I saw that you were perfect, and so I loved you. Then I saw that you were not perfect and I loved you even more", marriage: a}
+    fmt.Println("\n", l.quote, "\n")
+    //its possible to directly access the methods of embedded interface if the embedded interface is anonymous or unnamed.
+    l.love()
+    l.compatibility()
+    l.arranged()
+}
+
+```
+
+This example results in the following output:
+
+```shell
+
+love with john
+compatible with john
+arranged marriage with john
+
+ I saw that you were perfect, and so I loved you. Then I saw that you were not perfect and I loved you even more
+
+love with john
+compatible with john
+arranged marriage with john
+
+Program exited.
+
+```
+
+## Codebyte Example
+
+```codebyte/golang
```

Comment:
I think if we covered a simple interface in the previous examples and we want to include a code byte, we can display some extra functionality here.

Reply:
@yangc95  can you eloborate i didnt get this

Reply:
extra functionality hmm i didnt covered about few things like interface conversion,applied interfaces n stuff but if i just add code reg tht in codebyte without explaination it wont make sense...

what extra functionality are you referring to?

if you meant things like interface conversion n applied interfaces or anything more depth in interfaces i will make another PR after this as enhancement to interfaces and add things there with more explaination.

Reply:
Any enhancements would be great. As a base though, I think it would be good to set this concept up by breaking down how to create, and then implement interfaces. Within the create section, we can talk about syntax and give an example.

Then, in the implementation section, we can talk about how to implement an interface. Here, we can include the examples you've written so far with the addition of some meaningful explanation as to how the interface is being implemented. I think this is a better explanation.